ethereum/contracts/TablelandRigs.sol

```diff
@@ -354,6 +390,417 @@ contract TablelandRigs is
+            // i.e., `park` results in a status of `UNTRAINED` or `PARKED`
+            else return GarageStatus.PARKED;
+        } else {
+            // Invariant: pilot data cannot be `0` if `startTime` is > `0`
+            if (_pilotData(tokenId) == 1) return GarageStatus.TRAINING;
+            else return GarageStatus.PILOTED;
+        }
+    }
+
+    /**
+     * @dev Returns whether or not a Rig can be piloted.
+     *
+     * tokenId - the unique Rig token identifier
+     */
+    function _canPilot(uint256 tokenId) private view returns (bool) {
+        // Invariant: Cannot switch real pilots mid-flight, but can go from trainer -> real pilot
+        // There are two ways to pilot:
+        //   1. In a `PARKED` status
+        //   2. In a `TRAINING` status for long enough (30 days; 172800 blocks)
+        uint64 startTime = _startTime(uint16(tokenId));
+        if (startTime == 0) {
+            // Rig is either `UNTRAINED` or `PARKED`
+            // It's not possible to pilot from `UNTRAINED`
+            return _pilotData(uint16(tokenId)) != 0; // Could be a trainer (`1`) or real pilot
+        } else if (_pilotData(uint16(tokenId)) == 1) {
+            return block.number >= startTime + 172800; // Check training time completed
+        } else return false;
+    }
+
+    /**
+     * @dev See {ITablelandRigs-trainRig}.
+     */
+    function trainRig(uint256 tokenId) external {
+        // Check the Rig `tokenId` exists
+        if (!_exists(tokenId)) revert OwnerQueryForNonexistentToken();
+        // Verify `msg.sender` is authorized to train the specified Rig
+        if (ownerOf(tokenId) != _msgSenderERC721A()) revert Unauthorized();
+        // Validate the Rig is untrained
+        if (_pilotStatus(uint16(tokenId)) != GarageStatus.UNTRAINED)
+            revert InvalidPilotStatus();
+        // Assign a trainer pilot to the Rig in `_pilots`
+        // The "pilot data" is the pilot ID `1` and pilot contract `0`
+        _setStartTime(uint16(tokenId), uint64(block.number));
+        _setPilotData(uint16(tokenId), 0, 1);
+        // Insert the Rig training session into the Tableland pilot sessions table
+        TablelandDeployments.get().runSQL(
+            address(this),
+            _pilotSessionsTableId,
+            SQLHelpers.toInsert(
+                PILOT_SESSIONS_PREFIX,
+                _pilotSessionsTableId,
+                "rig_id,owner,start_time",
+                string.concat(
+                    StringsUpgradeable.toString(uint16(tokenId)),
+                    ",",
+                    SQLHelpers.quote(
+                        StringsUpgradeable.toHexString(ownerOf(tokenId))
+                    ),
+                    ",",
+                    StringsUpgradeable.toString(uint64(block.number))
+                )
+            )
+        );
+        emit Training(tokenId);
+    }
+
+    using ERC165Checker for address;
+
+    /**
+     * @dev See {ITablelandRigs-pilotRig}.
+     */
+    function pilotRig(
```

Comment:
@datadanne was thinking about the use case for pilot -> park -> pilot, where the same / most recent pilot is being used in the second piloting step.

with the second attempt, which would you expect to pass from the frontend:
1. `tokenId`, `pilotContract`, and `pilotId` -- i.e., pass the same values as the first piloting step
2. `tokenId`, `pilotContract` as `0x0`, and `pilotId` as `0`, and let the SC perform the necessary checks to pilot with the most recently used pilot
3. only `tokenId`, and let the SC perform the necessary checks

Reply:
Number 1! For 2 and 3 the client would have to keep a bunch of state and somehow recalculate that if the user reloads the page in between parking and piloting again